File: 3414.maximum-score-of-non-overlapping-intervals.py
#
# @lc app=leetcode id=3414 lang=python3
#
# [3414] Maximum Score of Non-overlapping Intervals
#

# @lc code=start
from bisect import bisect_right
from typing import List

class Solution:
    def maximumWeight(self, intervals: List[List[int]]) -> List[int]:
        # Add original index to intervals: [start, end, weight, original_index]
        # We use 0-based indexing for the intervals array internally.
        n = len(intervals)
        events = []
        for i in range(n):
            events.append(intervals[i] + [i])
        
        # Sort by end time. 
        # If end times are equal, sorting order doesn't strictly affect correctness of max weight,
        # but stable sort or specific secondary keys can make behavior deterministic.
        # We just sort by end time.
        events.sort(key=lambda x: x[1])
        
        # Precompute end times for binary search
        ends = [x[1] for x in events]
        
        # dp[k][i] stores the best (weight, negated_indices_tuple) 
        # choosing k intervals from the first i sorted intervals (events[0]...events[i-1]).
        # We use negated indices to leverage Python's default tuple comparison in max().
        # We want max weight, then min lexicographical indices.
        # Tuple comparison (w1, ids1) > (w2, ids2) checks w1 > w2, then ids1 > ids2.
        # If we store ids as (-i1, -i2, ...), then ids1 > ids2 corresponds to the lexicographically smaller set.
        
        # Initialize DP table
        # Rows 0 to 4 (representing 0 to 4 intervals chosen)
        # Cols 0 to n (representing prefix of sorted intervals)
        # Value: (weight, indices_tuple)
        # Initialize with a sentinel that is smaller than any valid outcome for k > 0.
        # For k=0, weight is 0, indices is ().
        
        dp = [[(-1, ())] * (n + 1) for _ in range(5)]
        for i in range(n + 1):
            dp[0][i] = (0, ())
            
        for i in range(1, n + 1):
            curr_interval = events[i-1]
            start, end, weight, original_idx = curr_interval
            
            # Find the index of the last interval that ends strictly before current start
            # bisect_right returns an insertion point p. 
            # All elements in ends[:p] are <= start - 1 < start.
            # So events[0...p-1] are compatible.
            # This corresponds to dp[...][p].
            p = bisect_right(ends, start - 1)
            
            for k in range(1, 5):
                # Option 1: Don't include current interval
                # Inherit from dp[k][i-1]
                best_score, best_indices = dp[k][i-1]
                
                # Option 2: Include current interval
                # We need to extend a valid solution of size k-1 from compatible intervals
                prev_score, prev_indices = dp[k-1][p]
                
                if prev_score != -1:
                    new_score = prev_score + weight
                    # Construct new indices tuple
                    # Convert negated indices back to positive to sort properly with new index
                    current_indices_list = sorted([-x for x in prev_indices] + [original_idx])
                    new_indices = tuple(-x for x in current_indices_list)
                    
                    # Compare
                    # Because we use negated indices, simple > comparison works for
                    # "MAX weight, then MIN lexicographical indices"
                    if (new_score, new_indices) > (best_score, best_indices):
                        best_score = new_score
                        best_indices = new_indices
                
                dp[k][i] = (best_score, best_indices)
        
        # The answer is the best among picking 1, 2, 3, or 4 intervals from all n intervals
        # i.e., max over k=1..4 of dp[k][n]
        final_res = (-1, ())
        for k in range(1, 5):
            if dp[k][n] > final_res:
                final_res = dp[k][n]
        
        # Convert negated indices back to original
        return sorted([-x for x in final_res[1]])

# @lc code=end

File: 3414.maximum-score-of-non-overlapping-intervals.cpp
#
# @lc app=leetcode id=3414 lang=cpp
#
# [3414] Maximum Score of Non-overlapping Intervals
#

# @lc code=start
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    struct State {
        long long score;
        int len;
        int idx[4];
    };

    static constexpr long long NEG = -(1LL<<60);

    static State invalidState() {
        State s;
        s.score = NEG;
        s.len = 0;
        for (int i = 0; i < 4; i++) s.idx[i] = 0;
        return s;
    }

    static bool lexLess(const State& a, const State& b) {
        int m = min(a.len, b.len);
        for (int i = 0; i < m; i++) {
            if (a.idx[i] != b.idx[i]) return a.idx[i] < b.idx[i];
        }
        return a.len < b.len; // shorter (prefix) is lexicographically smaller
    }

    static State better(const State& a, const State& b) {
        if (a.score == NEG) return b;
        if (b.score == NEG) return a;
        if (a.score != b.score) return (a.score > b.score) ? a : b;
        return lexLess(a, b) ? a : b;
    }

    static State addInterval(const State& base, int originalIdx, long long w) {
        if (base.score == NEG) return invalidState();
        State res = base;
        res.score += w;
        // insert originalIdx into res.idx keeping sorted
        int pos = res.len;
        while (pos > 0 && res.idx[pos - 1] > originalIdx) {
            res.idx[pos] = res.idx[pos - 1];
            pos--;
        }
        res.idx[pos] = originalIdx;
        res.len++;
        return res;
    }

    vector<int> maximumWeight(vector<vector<int>>& intervals) {
        int n = (int)intervals.size();
        struct Item { long long l, r, w; int idx; };
        vector<Item> a;
        a.reserve(n);
        for (int i = 0; i < n; i++) {
            a.push_back({(long long)intervals[i][0], (long long)intervals[i][1], (long long)intervals[i][2], i});
        }
        sort(a.begin(), a.end(), [](const Item& x, const Item& y) {
            if (x.r != y.r) return x.r < y.r;
            if (x.l != y.l) return x.l < y.l;
            return x.idx < y.idx;
        });

        // 1-based for convenience
        vector<long long> ends(n + 1, LLONG_MIN);
        vector<Item> items(n + 1);
        for (int i = 1; i <= n; i++) {
            items[i] = a[i - 1];
            ends[i] = items[i].r;
        }

        // dp[i][j]: best using first i intervals (sorted by r), choosing exactly j
        vector<array<State, 5>> dp(n + 1);
        for (int j = 0; j <= 4; j++) dp[0][j] = invalidState();
        dp[0][0].score = 0;
        dp[0][0].len = 0;

        for (int i = 1; i <= n; i++) {
            // find p(i): last index < i with end < l_i
            long long target = items[i].l - 1; // end <= l-1
            int p = (int)(upper_bound(ends.begin() + 1, ends.begin() + i, target) - ends.begin()) - 1;

            for (int j = 0; j <= 4; j++) {
                State bestSkip = dp[i - 1][j];
                State bestTake = invalidState();
                if (j > 0) {
                    bestTake = addInterval(dp[p][j - 1], items[i].idx, items[i].w);
                }
                dp[i][j] = better(bestSkip, bestTake);
            }
        }

        State ans = invalidState();
        for (int j = 0; j <= 4; j++) ans = better(ans, dp[n][j]);

        vector<int> res;
        res.reserve(ans.len);
        for (int i = 0; i < ans.len; i++) res.push_back(ans.idx[i]);
        return res;
    }
};
# @lc code=end
